Implemented framework for replacing servers

guglo opened the pull request:
1. Added servers landing page to show all the servers. Made it
interact with the menu.
2. Added context menu for the page and updated context menu
to be functioning
3. Updated Install wizard to deal with update page and refresh
update page whilt it is still in progress
4. Added replace server modal
5. Moved server details component to the components dir
6. Moved CollapsibleTable to the components dir
7. Added 3 placeholder pages to make sure the flow works

Note:
This is not fully tested, only tested a happy path.
There are lots of TODOs need to be resolved.

To test:

1. Test without available servers
you can just enter any ilo inputs

2. Test with available servers
need to go to day0 to get some manual servers
or discovered servers, they will  show up in the dropdown
list in replace detail modal

jeremy-moffitt commented on src/pages/UpdateServers.js:
```diff
@@ -0,0 +1,198 @@
+            </div>
+          </div>
+        </div>
+        <div className='wizard-content'>
```
one thing about using the wizard css classes is that its causing an overly large header and clips the scrolling awkwardly...  theres a couple of solutions, one that I tried successfully was to add another css class here that sets the height back to auto... it might make more sense to just not have things that arent part of the wizard be under "wizard-page" and "wizard-content" though if that creates more css issues than it solves it might not be worth it

jeremy-moffitt replied:
hmm it does mess with the margins and whatnot to remove those :-/ will push a patchset with the css class approach

jeremy-moffitt commented on src/styles/pages/ServerRoleSummary.less:
```diff
@@ -305,58 +305,4 @@
-      .material-icons {
-        padding-left: 0.3em;
-        vertical-align: middle;
-        color: @box-border-color;
```
uploaded a patchset to put this styling back in place

guglo replied:
saw you fixed it thanks.

jeremy-moffitt commented on src/InstallWizard.js:
```diff
@@ -63,31 +62,83 @@ class InstallWizard extends Component {
+    if(this.state.currentMenuName) {
+      let extraVars = ['currentMenuName', 'currentOperation', 'operationProps'];
+      this.persistedStateVars = this.persistedStateVars.concat(extraVars);
+    }
```
I'm a little confused about what this is doing, and what the impact may be of the persistedStateVars list being variable depending on other conditions.... so we're sometimes persisting 3 extra vars and sometimes not....

guglo replied:
those are the sets for global and persisted variables for install and update. Update need some extra vars.

jeremy-moffitt commented on src/InstallWizard.js:
```diff
@@ -178,7 +241,22 @@ class InstallWizard extends Component {
+    }
+    else {
+      //if it is update, it will use dynamically assembled this.state.pages when startUpdate is
+      //triggered, if it is install, it will use the static this.props.pages
```
if what is update? is this basically splitting between "install" and "configure" modes?

guglo replied:
I will use a const to indicate if it is update to make it a little clearer

jeremy-moffitt commented on src/InstallWizard.js:
```diff
@@ -274,6 +352,50 @@ class InstallWizard extends Component {
 
+  //this is triggered when start the update processes
+  //for example click repalce button on Replace server modal
+  startUpdate = (operation, pages, theProps) => {
```
its a little weird that this stuff is in the InstallWizard.js , I guess it allows for reusing the wizard functionality when doing a multi-step process as a result of a user choice, so its probably ok

guglo replied:
yeah...tried to reuse the code in InstallWizard. I will add more comments.

jeremy-moffitt commented on src/components/CollapsibleTable.js:
```diff
@@ -60,6 +66,68 @@ class CollapsibleTable extends Component {
     this.setState({showServerDetailsModal: true, activeRowData: rowData});
   }
 
+  handleShowMenuServerDetails = () => {
```
it occurs to me reading this that this component is very server list specific, but thats a pre-existing condition and not something introduced here

guglo replied:
see your point. maybe next round to cleanup.

jeremy-moffitt commented on src/components/PlaybookProcess.js:
```diff
@@ -506,7 +506,10 @@ class PlaybookProgress extends Component {
   componentDidMount() {
     this.globalPlaybookStatus = this.getGlobalPlaybookStatus();
-    this.processPlaybooks();
+    //TODO don't start process if passed in how to display commit error???
```
this comment isnt really clear

guglo replied:
will update

jeremy-moffitt commented on src/components/ReplaceServerDetails.js:
```diff
@@ -0,0 +1,321 @@
+// (c) Copyright 2017-2018 SUSE LLC
```
new file copyrights can just be 2018

guglo replied:
will update

jeremy-moffitt commented on src/pages/CloudDeployProgress.js:
```diff
@@ -162,7 +162,7 @@ class CloudDeployProgress extends BaseWizardPage {
           <PlaybookProgress
             updatePageStatus = {this.updatePageStatus} updateGlobalState = {this.props.updateGlobalState}
             playbookStatus = {this.props.playbookStatus}
-            steps = {PLAYBOOK_STEPS} deployConfig = {this.props.deployConfig}
```
is deployConfig no longer used?

guglo replied:
no it is not used in PlaybookProgress

jeremy-moffitt commented on src/pages/ServerRoleSummary.js:
```diff
@@ -28,10 +28,6 @@ class ServerRoleSummary extends BaseWizardPage {
-    this.state = {
-      autoServers: undefined,
-      manualServers: undefined
-    };
```
these are still referenced, is this just cleanup since they'll default to undefined?

guglo replied:
this is just clean up.

jeremy-moffitt commented on src/styles/components/collapsibletable.less:
```diff
@@ -0,0 +1,65 @@
+  .group-row {
+    border-bottom: 1px solid @box-border-color;
+    td {
+      background: linear-gradient(to bottom, @table-header-gradient-color 0%, @table-header-color 100%);
```
this is cool :)

jeremy-moffitt commented on src/styles/components/contextmenu.less:
```diff
@@ -17,12 +17,9 @@
   position: absolute;
   z-index: 999;
-  width: 120px;
+  width: 103px;
```
wondering if this will be enough space if we get some longer action names...

guglo replied:
will make it a little bigger

jeremy-moffitt commented on src/utils/DynamicPagesFinder.js:
```diff
@@ -0,0 +1,35 @@
+  'servers_ReplaceServer': UpdateServerPages //<menu>_<operation> : <all the possible pages defined>
+};
+
+export function getPages(menuName, operationName, steps) {
```
its not at all clear what this class or function are doing... needs comments

guglo replied:
will update

jeremy-moffitt reviewed (changes requested):
generally speaking looks good... theres some comments I made outside of the "review" (yes, I should've just started a review) that need a look... please take a look at those

GarySmith commented on src/components/ReplaceServerDetails.js:
```diff
@@ -0,0 +1,321 @@
+        this.allInputsStatus[key] = INPUT_STATUS.INVALID;
+      }
+    }
+    this.setState({isFormValid: this.isFormTextInputValid()});
```
remove

guglo replied:
will update

GarySmith commented on src/components/ReplaceServerDetails.js:
```diff
@@ -0,0 +1,321 @@
+  handleOsPasswordChange = (e) => {
+    const password = e.target.value;
+    this.setState({osInstallPassword: password}, () => {
+      this.setState({isFormValid: this.isFormTextInputValid()});
```
Remove. 
Not only is this unnecessary (as mentioned above), this code is actually wrong. this.setState executes in a promise, and when that promise returns, then rendering can begin. But the callback here is another call to setState which immediately returns a promise, so `isFormValid` is likely to not be set until *after* rendering, which is wrong.

guglo replied:
will update to simplify and clean up

GarySmith commented on src/components/ReplaceServerDetails.js:
```diff
@@ -0,0 +1,321 @@
+  handleInstallOsCheck = () => {
+    this.setState({
+      isInstallOsChecked: !this.state.isInstallOsChecked }, () => {
+      this.setState({isFormValid: this.isFormTextInputValid()});
```
Ditto above. Remove this

guglo replied:
k

GarySmith commented on src/components/ReplaceServerDetails.js:
```diff
@@ -0,0 +1,321 @@
+        <ActionButton type='default'
+          clickAction={this.handleCancel} displayLabel={translate('cancel')}/>
+        <ActionButton
+          isDisabled={!this.state.isFormValid}
```
Replace the reference to `this.state.isFormValid` with a call to `this.isFormTextInputValid`

guglo replied:
k

GarySmith commented on src/components/ReplaceServerDetails.js:
```diff
@@ -0,0 +1,321 @@
+  constructor(props) {
+    super(props);
+
+    this.allInputsStatus = {
```
These status fields should be part of state, since they directly affect what is being rendered

guglo replied:
will see what I can do

GarySmith commented on src/components/ReplaceServerDetails.js:
```diff
@@ -0,0 +1,321 @@
+    let theServer = allAvailableServers.find(server => server.id === serverId);
+    let newJSData = {};
+    if(theServer) {
+      newJSData['ilo-ip'] = theServer['ilo-ip'] ? theServer['ilo-ip'] : '';
```
While this works, it is more straightforward to use:
```
newJSData['ilo-ip'] = theServer['ilo-ip'] || '';
```
on this line and the ones that follow

guglo replied:
will update

GarySmith commented on src/components/ReplaceServerDetails.js:
```diff
@@ -0,0 +1,321 @@
+      <InputLine
+        isRequired={this.state.isInstallOsChecked} inputName={'osInstallPassword'}
+        inputType={'password'} label={'server.pass.prompt'}
+        inputValue={this.state.osInstallPassword ? this.state.osInstallPassword : ''}
```
use `this.state.osInstallPassword || ''`

guglo replied:
will update

GarySmith commented on src/pages/ReplaceServer/PrepareReplace.js:
```diff
@@ -0,0 +1,123 @@
+    }
+  }
+
+  goForward = (e) => {
```
Is this function needed? It should just call the parent's goForward without this, right?

guglo replied:
will update

GarySmith commented on src/pages/UpdateServers.js:
```diff
@@ -0,0 +1,198 @@
+    //deal with timing issue when load model
+    this.setState({
+      model : newProps.model,
+    } , () => {
```
As mentioned above, calling `this.setState` in the callback to `this.setState` is problematic and should be avoided.

guglo replied:
will update

GarySmith reviewed (changes requested):
This is great work, and I was amazed to see how functional it is already. Just need to address some stuff.

guglo commented:
@GarySmith @jeremy-moffitt  thanks for the review, I will go through the comments and address them when I get back to the further development.

guglo commented on src/InstallWizard.js:
```diff
@@ -63,31 +62,83 @@ class InstallWizard extends Component {
+      //deal with update
+      currentMenuName: this.props.menuName,
+      currentOperation: undefined,
+      pages: undefined,
```
I will add comments to explain

guglo commented on src/InstallWizard.js:
```diff
@@ -63,31 +62,83 @@ class InstallWizard extends Component {
       deployConfig: undefined, // cloud deployment configuration
+
+      //deal with update
+      currentMenuName: this.props.menuName,
```
yeah...i will add a const to indicate if it is update. The goal is to reuse the code in InstallWaizard.

guglo commented on src/InstallWizard.js:
```diff
@@ -126,6 +172,20 @@ class InstallWizard extends Component {
       });
   }
 
+  componentWillMount = () => {
```
componentWillMount is for update and componentDidMount is for install. loading in componentDidMount for update has some issues. so leave the install load in componentDidMount as it is.  I will do some more testing to see if I can move update into componentDidMount

jeremy-moffitt replied:
was there any luck with this?

jeremy-moffitt replied:
I missed the "if" statement with the IS_UPDATE conditional... I am curious about why it works differently, but at least its not loading the model twice

guglo replied:
Tried a couple of times today...wired, the problem seems go away...will keep the loading in the original componentDidMount.

guglo commented on src/InstallWizard.js:
```diff
@@ -63,31 +62,83 @@ class InstallWizard extends Component {
     // To simulate a delay in startup and to be able to see the initial loading mask,
-    //    uncomment the following line and the line with the closing parenthesis two lines later
-    // new Promise(resolve => setTimeout(resolve, 1000)).then(() =>
+    // uncomment the following line and the line with the closing parenthesis two lines later
```
will update

guglo commented on src/InstallWizard.js:
```diff
@@ -126,6 +172,20 @@ class InstallWizard extends Component {
+  componentWillMount = () => {
+    //update
+    if(this.props.menuName) {
+      this.loadModelandProgress(this.loadUpdateProgress);
```
will update

guglo commented on src/InstallWizard.js:
```diff
@@ -274,6 +352,50 @@ class InstallWizard extends Component {
+    //build steps dynamically
+    let steps = pages.map(page => { return {'name': page.name}; });
+    //get the first one start
+    steps[0].stepProgress = 2; //in progress
```
will udpate

guglo commented on src/components/CollapsibleTable.js:
```diff
@@ -97,7 +165,7 @@ class CollapsibleTable extends Component {
         'groupName': g,
         'members': groupMap.get(g),
-        'isExpanded': this.props.expandedGroup.includes(g)}))
+        'isExpanded': this.props.expandedGroup ? this.props.expandedGroup.includes(g) : false}))
```
will update

guglo commented on src/components/CollapsibleTable.js:
```diff
@@ -202,14 +292,41 @@ class CollapsibleTable extends Component {
 
+  renderReplaceServerModal() {
+    let title =
+      translate('server.replace.heading') + (this.state.activeRowData ? '- ' +
```
will update

guglo commented on src/components/ContextMenu.js:
```diff
@@ -58,16 +23,14 @@ class ContextMenu extends Component {
 
   componentDidMount() {
-    document.addEventListener('click', this.handleClickOutside);
+    document.addEventListener('click', this.handleCloseMenu);
```
this handles click outside menu area. when user clicks menu outside area, the menu closes. will update comments

guglo commented on src/components/ReplaceServerDetails.js:
```diff
@@ -0,0 +1,321 @@
+    };
+
+    this.state = {
+      isFormValid: false,
```
yeah...will update

guglo commented on src/components/ReplaceServerDetails.js:
```diff
@@ -0,0 +1,321 @@
+      }
+      else {
+        this.setState({osInstallPassword: value}, () => {
+          this.setState({isFormValid: this.isFormTextInputValid()});
```
will update

guglo commented on src/components/ReplaceServerDetails.js:
```diff
@@ -0,0 +1,321 @@
+
+  updateFormValidity = (props, isValid) => {
+    this.allInputsStatus[props.inputName] = isValid ? INPUT_STATUS.VALID : INPUT_STATUS.INVALID;
+    this.setState({isFormValid: this.isFormTextInputValid()});
```
will update

GarySmith commented:
The approach of folding in the update functionality into the existing InstallWizard seems problematic; the InstallWizard is built around the assumption that only one user can be running the application at a time. In day2 scenarios, it is possible that more than one user is logged in at the same time, so this is going to be problematic if this wizard is used at the same time by different users (for example two unrelated. If we are going to merge this in its current state, then we should add a new story to the backlog to refactor this.

GarySmith commented on src/components/ModelServerDetails.js:
```diff
@@ -13,7 +13,8 @@
 import React, { Component } from 'react';
-import { translate } from '../../localization/localize.js';
+import { translate } from '../localization/localize.js';
+import { maskPassword } from '../utils/ModelUtils.js';
```
It does not make sense to import this `maskPassword` function and then overwrite it on line 33.

guglo replied:
ah...function at line 33 was moved the ModelUtils, supposed to be removed. Will remove it here.

GarySmith commented on src/pages/BaseWizardPage.js:
```diff
@@ -73,7 +74,17 @@ class BaseWizardPage extends Component {
   }
 
-  renderNavButtons() {
+  closeUpdateProcess(e) {
```
These functions that are specific to the update do not belong in this class, which is a base class for any page that appear in the wizard.

guglo replied:
All the update pages reused most of the components from the BaseWizardPage,  Do you mean I need to have a another say BaseUpdateWizardPage which extends from BaseWizardPage to have closeUpdateProcess and cancelUpdateProcess moved there?

GarySmith replied:
Yes, exactly. The base class should only contain things that are common to all pages

jeremy-moffitt commented:
I dont think the ardana framework supports parallel updates... it does pose an interesting question for the multi-user scenario, I guess we'd have to lock change functionality while a change is going on if we're not assuming that any user logged in during an update is party to the update

GarySmith commented on src/utils/InputValidators.js:
```diff
@@ -63,6 +79,21 @@ export function MacAddressValidator(macAddress) {
+  if(props && props.exist_availservers_mac_addr_objs) {
+    let find =
+      props.exist_availservers_mac_addr_objs.find(mac => mac['mac-addr'] === macAddress);
+    if (find) {
```
This logic is broken. `find()` returns the array element, so if the first element (`[0]`) were a match, this logic would not detect it. You should probably replace `find()` with `some()`, which returns `true` if there is a match (and also rename the `find` variable to something like `found`)

guglo replied:
my element is an array of objects.  I can not use some() since I want the object back so I can use the serverId to construct my message. I think the better way is to change if(find) to be if (found !== undefined)

GarySmith replied:
You are right, it was my misunderstanding. The code here is fine.

GarySmith commented on src/utils/constants.js:
```diff
@@ -36,6 +36,10 @@ export const MODEL_SERVER_PROPS =  [
   'id', 'ip-addr', 'mac-addr' ,'server-group' ,'nic-mapping' ,'ilo-ip' ,'ilo-user', 'ilo-password'
 ];
 
+export const REPlACE_SERVER_PROPS =  [
```
Typo here (the l in REPlACE should be in uppercase)

guglo replied:
will update

GarySmith commented:
@jeremy-moffitt I would not necessarily rule out concurrent updates (since concurrent playbook runs are certainly allowed). But regardless of that, the install wizard was designed so that if a second user happens to connect (of the single user re-connects), then they are both shown the same screen. I don't think that we would want that behavior in a multi-user UI.

guglo commented on src/utils/DynamicPagesFinder.js:
```diff
@@ -0,0 +1,49 @@
+ * @param steps              the steps for the process.
+ * @returns pages            assembled page components based on steps
+ */
+export function getPages(menuName, operationName, steps) {
```
getPages is called from InstallWizard During loadUpdateProgress.  It uses menuName, operation and steps to determine what pages are needed to load. Even UpdateServerPages are specific for the update servers, need a common place where wizard can pick up what is the left.

guglo commented:
@GarySmith @jeremy-moffitt  regarding multi users ...assume we don't allow concurrent updates. However at this point, no restrictions. Need to have a restriction when update is in progress.

guglo commented:
I agree that we need have a design for how to address multiple administrators access and perform operations in  the day2UI. I think it might be better to address multiple users in the separate commit and refactor the code if we need to. At this point, I will record the current user name once user kicks off the update process so it could be used to restrict pages for readonly purpose for other users later.

Commit:
Rebased with latest master and added a summary tab

Implemented framework for replacing servers

1. Added servers landing page to show all the servers. Made it
   interact with the menu.
2. Added context menu for the page and updated context menu
   to be functioning
3. Updated Install wizard to deal with update page and refresh
   update page whilt it is still in progress
4. Added replace server modal
5. Moved server details component to the components dir
6. Moved CollapsibleTable to the components dir
7. Added 3 placeholder pages to make sure the flow works

Note:
There are lots of TODOs need to be resolved.
Select from available servers need further work.
There is no restriction for multiple users running the update,
need to think about the design to restrict the funcitonalities
when updating is in progress.
The progress.json need to be cleaned before start day2 UI

To test:
1. Test without available servers
   you can just enter any ilo inputs

2. Test with available servers
   need to go to day0 to get some manual servers
   or discovered servers, they will  show up in the dropdown
   list in replace detail modal

GarySmith commented:
This appear to address the concerns I had and to implement the improved design that Gloria and I discussed last week. Since github cannot gracefully handle forced pushes in reviews, it is nearly impossible to tell exactly what changed between this version and the previous one.
